### PR TITLE
Modelize type intersections

### PR DESCRIPTION
**Pull request by jcbrinfo**
Add `MIntersectionType` to the model and use this new kind of type to enhance adaptative typing. As a side-effect, do some refactoring concerning the detection of nullables and primitive types.

For now, only essential (to avoid regressions) simplifications of type expressions are implemented.

Note: `MTypeSet` is designed to become the common superclass of `MIntersectionType` ( type intersections) and `MUnionType` (type unions) in future PRs.

### CODE REVIEW

**Comment by jcbrinfo** [comment]
[#INF7845](http://web.archive.org/web/20170326053414/http://info.uqam.ca/~privat/INF7845/travail_de_session/)

**Comment by privat** [comment]
This is quite nice. I think I need more time to review (and maybe discuss) the intersection API and implementation. However, I think the initial cleanup (and factorized maybenull) are useful alone. I suggests that you PR the cleanup part independently so it can be merged first (and faster). This will also simplify the review of the current PR.

**Comment by jcbrinfo** [comment]
♻️ Fix `nitcs` and `nitce`.

**Comment by jcbrinfo** [comment]
♻️ Take 3: Fix `nitj`.
I totally forgot that I had to compile that one separately. Sorry for the mess.

**Comment by jcbrinfo** [comment]
♻️ Add `MIntersectionType::hash`

File: src/model/model.nit [hunk 1]
```diff
@@ -1095,6 +1113,46 @@ abstract class MType
...
+	fun can_be_null(mmodule: MModule, anchor: nullable MClassType): Bool
+	do
+		return false
+	end
+
+	# Intersect type with the specified type.
+	#
+	# The resulting type represents the subtypes that are common to both `self`
+	# and `other`. The contextual parameters (`mmodule` and `anchor`) are
+	# used to simplify the resulting type.
```
Comment (jcbrinfo) [inline src/model/model.nit]: Make types comparable in order to simplify.

File: src/model/model.nit [hunk 2]
```diff
@@ -1188,6 +1248,287 @@ abstract class MType
...
+abstract class MTypeSet[E: MType]
+	super MType
+
+	# The module of reference.
+	#
+	# Used to populate `model`, find primitive types and internally build new
+	# types with the same operator.
+	#
+	# SEE: `apply_to`
+	var mmodule: MModule
```
Comment (jcbrinfo) [inline src/model/model.nit]: * When comparing types and when adding `Object`.
* `apply_to` does use this attribute in all its concrete implementations.

File: src/model/model.nit [hunk 3]
```diff
@@ -1188,6 +1248,287 @@ abstract class MType
...
+	fun apply_to(operands: Set[MType], mmodule: MModule,
+			anchor: nullable MClassType): MType is abstract
+
+	# Return the cached version of this operation over the specified set of operands.
+	#
+	# Intended to be used internally to cache results of methods like
+	# `as_notnull`, `lookup_fixed`, `resolve_for` and `undecorate`.
+	#
+	# SEE: `apply_to`
+	protected fun cache(operands: Set[MType]): MType
```
Comment (jcbrinfo) [inline src/model/model.nit]: I cache what was already cached in other `MType` classes.

File: src/model/model.nit [hunk 4]
```diff
@@ -1188,6 +1248,287 @@ abstract class MType
...
+#
+# WARNING: Don’t instantiate this class directly: use `MType::intersection`
+# instead. Else, you may end up with undefined behaviors.
+class MIntersectionType
+	super MTypeSet[MType]
+
+	# The intersected types.
+	#
+	# Must not be empty. Must not be edited.
+	fun constraints: Set[MType] do return operands
```
Comment (jcbrinfo) [inline src/model/model.nit]: Dates back to the time where `MTypeSet` did not exist. I may rename all calls to it if you want.

File: src/compiler/separate_erasure_compiler.nit [hunk 5]
```diff
@@ -516,12 +529,17 @@ class SeparateErasureCompilerVisitor
...
+		if mtype isa MNotNullType then
+			mtype = mtype.mtype
+		end
 		if mtype isa MParameterType then
 			# Here we get the bound of the the formal type (eh, erasure...)
 			mtype = mtype.resolve_for(self.frame.mpropdef.mclassdef.bound_mtype, self.frame.mpropdef.mclassdef.bound_mtype, self.frame.mpropdef.mclassdef.mmodule, false)
 			if mtype isa MNullableType then
 				mtype = mtype.mtype
-				accept_null = "1"
+				if maybe_null then
```
Comment (jcbrinfo) [inline src/compiler/separate_erasure_compiler.nit]: If it is wrapped in a type that forbids `null`.

File: src/model/model.nit (hunk 3 above)
Comment (jcbrinfo) [inline src/model/model.nit]: Also, caching is a trade-off between hash table accesses and the cost of `==` between `MType`s (from what I understand from the rest of the model). I hesitated to add the cache at the level of `MModule` because when intersection will be flattened, there may be a lot of temporary types created just to simplify the algorithms. For example, to compute a flattened version of `A and B and C`, we may create `A and B` first, then discard it. If only we had weak references, the choice would be easy.

**Comment by jcbrinfo** [comment]
@xymus I agree for the “TODOs” commit. But, for the 3 other commits that impact `MType::intersection`, the first is the basic definition, while the other are about moving code from `typing`. I could split everything in 3 PRs, but I thought it would be harder to review.

File: src/model/model.nit (hunk 2 above)
Comment (xymus) [inline src/model/model.nit]: OK, I understand how/where it is used but which mmodule is it? or is it any module? Other types are not linked to a module so I honestly don't know.

Not to insist but the only one implementation of  `apply_to` I see in this PR doesn't use it. However, it does a parameter of the same name.

File: src/model/model.nit (hunk 1 above)
Comment (xymus) [inline src/model/model.nit]: Right OK, they are used to resolve in order to simplify. The doc mention optimizing types somewhere else in this PR, is "simplify" and "optimize" the same concept?

**Comment by xymus** [comment]
About the commits, I don't want you to change this PR. But I do have to insist that breaking up a single method across multiple commits does not help the review process. I tried to read commit by commit but the code was not coherent in each commit, there were useless variables and the doc described features not yet present. You must try to create commits with bigger commits encompassing larger features.

File: src/model/model.nit (hunk 1 above)
Comment (jcbrinfo) [inline src/model/model.nit]: Yes.

Comment (jcbrinfo) [inline src/model/model.nit]: May be I should just use “simplify” everywhere.

**Comment by jcbrinfo** [comment]
♻️ 
- [X] Reformat `SeparateCompilerVisitor::autobox`
- [X] Rework the documentation of `MType::can_be_null`
- [X] Rework the documentation of `MType::intersection`
- [X] Rework the documentation of `MTypeSet::mmodule`
- [X] Rework the documentation of `MTypeSet::operands`
- [X] Rework the documentation of `MTypeSet::separator`
- [X] Remove attribute `MIntersectionType::constraints`
	- [X] Rename `MIntersectionType::from_constraints` to `from_operands`
- [X] Move the by-operands cache to the module.
- [X] Reformat `MParameterType::resolve_for_anchored`

**Comment by jcbrinfo** [comment]
♻️ Fix `nitj`

**Comment by jcbrinfo** [comment]
Fails seem to be related to unavailable ports on the test runner.

> libevent warning: Opening localhost:18946 failed

File: src/model/model.nit [hunk 6]
```diff
@@ -1201,6 +1259,289 @@ abstract class MType
...
+	#
+	# SEE: `cache`
+	protected fun by_operands_cache(mmodule: MModule): Map[Set[MType], MType] is abstract
+
+	# Return the cached version of this operation over the specified set of operands.
+	#
+	# `mmodule` is the module in which the cache resides.
+	#
+	# Intended to be used internally to cache results of methods like
+	# `as_notnull`, `lookup_fixed`, `resolve_for` and `undecorate`.
```
Comment (xymus) [inline src/model/model.nit]: Continuing the discussion from #2402. In Nit `protected` doesn't have the same meaning than in Java. `model` is a self-containing module, its clients are the engines and they don't define new types. If the clients/engines don't need access to the cache, this method should be private.

**Comment by jcbrinfo** [comment]
:recycle:
- [x] Promote resolution logic to `MTypeSet`.
- [x] Order definitions in `MIntersectionType` by name.
- [x] Move cache to `MType` so any caller of `intersection` benefit from it and it no longer use a set as a key.
- [x] Cache `hash`, so retrieving an intersection of intersections from the cache is faster.
- [x] Allow null anchors for `MFormalType::can_be_null` so `nitmetrics` can compile.
- [x] Remove an unecessary redefiniton of `can_be_null`.

Here are the Ir counts for `(cd src && ./valgrind.sh -- ../bin/nitc -o /dev/null "$module")`:

| Cache for intersections | Cache for `hash`  | `nitc.nit`<sup>1</sup> | `../tests/alt/base_adaptive_full_alt2.nit` |
| ---  | --- | ---: | ---------------: |
| No                      | No                | 10 628 637 825         | 53 062 955                                 |
| Yes                     | No                | 10 624 738 666         | 52 798 002                                 |
| Yes                     | Yes               | 10 618 383 914         | 52 798 022                                 |

Thank you @xymus for your precious feedback.

<hr/>

<sup>1</sup> The current version of this PR.

File: src/model/model.nit [hunk 7]
```diff
@@ -1201,6 +1273,265 @@ abstract class MType
...
+	# Apply the same operator over `operands`.
+	#
+	# Does not consider properties of `self`. Used to abstract type
+	# generation done by methods of this class.
+	#
+	# May simplify the resulting expression and not return an instance of
+	# `SELF`. The contextual parameters (`mmodule` and `anchor`) are used to
+	# resolve types during comparisons in order to simplify the resulting type.
+	protected fun apply_to(operands: Set[MType], mmodule: MModule,
+			anchor: nullable MClassType): MType is abstract
```
Comment (xymus) [inline src/model/model.nit]: I think that I left this comment before, but it may have been on another service.

This method must be made private. Its only purpose is to factorize code in the same module and the only expected subclasses are in the same module. As such, it is not useful to the clients of the model API and it should be private to the module.

Comment (jcbrinfo) [inline src/model/model.nit]: The problem I have with making this method private is that it indirectly makes `MTypeSet` “final” for all the other modules. Since `apply_to` is abstract, any subclass must redefine it. So, it makes sense to me to make this method visible to any subclass.

**Comment by jcbrinfo** [comment]
♻️ 
- [x] Make `MTypeSet::apply_to` redefinable only in `model`.

File: src/model/model.nit [hunk 8]
```diff
@@ -1201,6 +1273,265 @@ abstract class MType
...
+		var result = i.item
+		i.next
+		for operand in i do
+			result = result.intersection(operand, mmodule, anchor)
+		end
+		return result
+	end
+
+	redef var as_notnull is lazy do
+		var new_operands = new Set[MType]
```
Comment (privat) [inline src/model/model.nit]: I'm not sure to understand why you need such a complex approach?
Why not applying `as_notnull` on each element then `apply_to` them?

File: src/model/model.nit [hunk 9]
```diff
@@ -1201,6 +1273,265 @@ abstract class MType
...
+# A type that represents an operation over a set of types.
+abstract class MTypeSet[E: MType]
+	super MType
+
+	# The context in which this type was created.
+	#
+	# Used to populate `model` and look for `Object`.
+	#
+	# TODO: Remove this attribute once `as_notnull` don’t need it.
+	var mmodule: MModule
```
Comment (privat) [inline src/model/model.nit]: You do not need it in `MTypeSet` and I'm not sure you really need it in `MIntersectionType`

File: src/model/model.nit [hunk 10]
```diff
@@ -1201,6 +1273,265 @@ abstract class MType
...
+
+	# Apply the same operator over `operands`.
+	#
+	# Does not consider properties of `self`. Used to abstract type
+	# generation done by methods of this class.
+	#
+	# May simplify the resulting expression and not return an instance of
+	# `SELF`. The contextual parameters (`mmodule` and `anchor`) are used to
+	# resolve types during comparisons in order to simplify the resulting type.
+	private fun apply_to(operands: Set[MType], mmodule: MModule,
```
Comment (privat) [inline src/model/model.nit]: I think that `mmodule` should be nullable since it prevents to run apply_to globally, ie without a specific module as context. This may be related the present of your unneeded `mmodule` attribute

File: src/model/model.nit [hunk 11]
```diff
@@ -1201,6 +1273,265 @@ abstract class MType
...
+				result.add_all(mtype.collect_mtypes(mmodule))
+			end
+			collect_mtypes_cache[mmodule] = result
+		end
+		return result
+	end
+
+	private var collect_mtypes_cache = new Map[MModule, Set[MClassType]]
+
+	redef var hash is lazy do return operands.hash
```
Comment (privat) [inline src/model/model.nit]: I like when `hash` is near `==` so one can see that `hash` was not forgotten and that the laws of the univers are enforced.

File: src/model/model.nit [hunk 12]
```diff
@@ -1201,6 +1273,265 @@ abstract class MType
...
+
+	redef var undecorate is lazy do
```
Comment (privat) [inline src/model/model.nit]: why this one is not in `MTypeSet`?

File: tests/base_adaptive_intersection_formal.nit [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Test of intersections with a formal type in the context of the adaptive typing.
+
+class Foo
```
Comment (privat) [inline tests/base_adaptive_intersection_formal.nit]: `import core::kernel` make the tests faster

File: src/compiler/separate_erasure_compiler.nit (hunk 5 above)
Comment (privat) [inline src/compiler/separate_erasure_compiler.nit]: This seems unrelated to the commit; is there a relation I do miss?

File: tests/sav/base_adaptive_full_alt2.res [hunk 14]
```diff
@@ -1,16 +1,7 @@
-alt/base_adaptive_full_alt2.nit:55,10: Type Error: expected `B`, got `C`.
-alt/base_adaptive_full_alt2.nit:71,10: Type Error: expected `B`, got `C`.
-alt/base_adaptive_full_alt2.nit:84,11: Type Error: expected `B`, got `C`.
-alt/base_adaptive_full_alt2.nit:91,10: Type Error: expected `A`, got `Object`.
-alt/base_adaptive_full_alt2.nit:95,10: Type Error: expected `B`, got `Object`.
```
Comment (privat) [inline tests/sav/base_adaptive_full_alt2.res]: To improve the proof of great success, uncomment the successful `#alt2#` from `base_adaptive_full.nit`.
You should also update the module comment.

File: src/model/model.nit [hunk 15]
```diff
@@ -1201,6 +1273,265 @@ abstract class MType
...
+		end
+		return result
+	end
+
+	redef var full_name is lazy do
+		var names = new Array[String].with_capacity(operands.length)
+		for mtype in operands do
+			names.add(mtype.full_name)
+		end
+		return names.join(separator)
```
Comment (jcbrinfo) [inline src/model/model.nit]: It’s left for a future PR. ;)

But, I can merge with #2402 if you want.

File: src/model/model.nit (hunk 8 above)
Comment (jcbrinfo) [inline src/model/model.nit]: Because the days of `NotNullType` are counted.

File: src/model/model.nit (hunk 11 above)
Comment (jcbrinfo) [inline src/model/model.nit]: I was hesitating between that and keeping the lexicographic order.

File: src/model/test_model.nit [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+
+	init do
+		model = new Model
+		var location = new Location(null, 1, 1, 1, 1)
+		mmodule = new MModule(model, null, "module_foo", location)
+		object_type = create_class("Object", interface_kind,
+				new Array[MClassType])
+		a_type = create_class("A", concrete_kind, [object_type])
+		b_type = create_class("B", concrete_kind, [object_type])
+		d_type = create_class("D", concrete_kind, [a_type, b_type])
```
Comment (jcbrinfo) [inline src/model/test_model.nit]: It was created before ModelDiamond. That’s already legacy code! :)

File: src/model/model.nit (hunk 9 above)
Comment (jcbrinfo) [inline src/model/model.nit]: My idea is to move it as a parameter in a future PR (already done, just not submitted). The only alternative I see is to change the way `Object` is introduced in the model.

File: src/model/model.nit (hunk 12 above)
Comment (jcbrinfo) [inline src/model/model.nit]: Because I forgot obviously. ;)

File: src/compiler/separate_erasure_compiler.nit (hunk 5 above)
Comment (jcbrinfo) [inline src/compiler/separate_erasure_compiler.nit]: See my discussion about this with @xymus above. It’s a problem that occurs only since we have intersections.

File: src/model/model.nit (hunk 10 above)
Comment (jcbrinfo) [inline src/model/model.nit]: It’s only an indirection for `intersection` (or `union`). Since `is_subtype` requires the module, so do `intersection` and `apply_to`.

File: src/model/model.nit (hunk 12 above)
Comment (jcbrinfo) [inline src/model/model.nit]: Haha! Now I remember why: it’s because the implementation will be totally different for unions when they will replace `NullabeType` (simply remove the `NullType` from the operands). I thought by keeping `undecorate` abstract for `MTypeSet`, at least I won’t forgot to redefine it for unions.

File: src/model/model.nit (hunk 8 above)
Comment (jcbrinfo) [inline src/model/model.nit]: More precisely, `T.as_not_null == T and Object`. However, because `Object` denies `null`, `intersection` would call `T.as_not_null` and… we get caught in an infinite recursion.

Comment (jcbrinfo) [inline src/model/model.nit]: Now that I told you this, I notice that I made a big mistake at the last line.

**Comment by jcbrinfo** [comment]
:recycle:
- [x] Only compute the depth of each operand once.
- [x] Add a to-do for ambiguous string representations.
- [x] Move most of the implementation of `undecorate` to `MTypeSet` (and add a to-do to not forgot to override it for unions).
- [x] Move `hash` near `==`.
- [x] Avoid infinite recursion in `as_notnull`
- [x] Add a test that provokes an infinite recursion if `intersection` and `as_notnull` are mutually recursive.
- [x] Import only `core::kernel` in `test/base_adaptive_intersection_formal.nit`
- [x] Use `ModelDiamond`.
- [x] Remove `#alt2#` for now allowed code in `tests/base_adaptive_full.nit`
- [x] Commit the implementation of `apply_to` and `as_notnull` separately from the rest of `MIntersectionType` so the compiler is semantically valid at each commit.
- [x] Flatten the intersections so we can compare them in the tests.
- [x] Allow any collection of types for `apply_to` (no need to require a set).

Remaining discussion points:
* The `mmodule` attribute.
* The signature of `apply_to`.
* `full_name`
* The implementation of `as_notnull`.
* Whether or not some changes should appear in the commit about `nitce`.
